engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java

```diff
@@ -249,15 +252,10 @@ private void initRenderGraph() {
         renderGraph.add(bloomPassesNode);
         renderGraph.add(blurPassesNode);
         renderGraph.add(finalPostProcessingNode);
-    }
 
-    private static <T extends Node> T createInstance(Class<T> type, Context context) {
-        // Attempt constructor-based injection first
-        T node = InjectionHelper.createWithConstructorInjection(type, context);
-        // Then fill @In fields
-        InjectionHelper.inject(node, context);
-        node.initialise();
-        return type.cast(node);
+        RenderPipelineGenerator generator = new RenderPipelineGenerator(renderGraph);
+        RenderPipelineOptimizer optimizer = new RenderPipelineOptimizer(generator.generate());
+        renderPipelineExecutor = new RenderPipelineExecutor(optimizer.optimize());
```

Comment:
The problem with this code is that it obfuscate the assembly-chain-like relationship between the various objects.  We can leave it as it is for the time being. I suspect we'll need some kind of Manager class though, calling:

orderedNodes = renderGraph.getNodesInTopologicalOrder();
renderPipeline = RenderPipelineGenerator.createPipelineFrom(orderedNodes);
optimizedPipeline = RenderPipelineOptimizer.optimize(renderPipeline);

and eventually

optimizedPipeline.execute();

In this way you can clearly see the way the data is being processed. Also, this would simplify a little the generator and optimizer classes. And a manager might be needed anyway to refresh these lists depending on changes to the dag. But as I said, let's leave these lines as they are for the time being.